// @ledgerhq/hw-transport-node-hid/lib/TransportNodeHid.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var node_hid_1 = __importDefault(require("node-hid"));
var hw_transport_node_hid_noevents_1 = __importStar(require("@ledgerhq/hw-transport-node-hid-noevents"));
var devices_1 = require("@ledgerhq/devices");
var errors_1 = require("@ledgerhq/errors");
var listenDevices_1 = __importDefault(require("./listenDevices"));
var listenDevicesDebounce = 500;
var listenDevicesPollingSkip = function () { return false; };
/**
 * node-hid Transport implementation
 * @example
 * import TransportNodeHid from "@ledgerhq/hw-transport-node-hid";
 * ...
 * TransportNodeHid.create().then(transport => ...)
 */
var TransportNodeHid = /** @class */ (function (_super) {
    __extends(TransportNodeHid, _super);
    function TransportNodeHid() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * if path="" is not provided, the library will take the first device
     */
    TransportNodeHid.open = function (path) {
        return Promise.resolve().then(function () {
            if (path) {
                return new TransportNodeHid(new node_hid_1["default"].HID(path));
            }
            var device = (0, hw_transport_node_hid_noevents_1.getDevices)()[0];
            if (!device)
                throw new errors_1.TransportError("NoDevice", "NoDevice");
            return new TransportNodeHid(new node_hid_1["default"].HID(device.path));
        });
    };
    /**
     *
     */
    TransportNodeHid.isSupported = hw_transport_node_hid_noevents_1["default"].isSupported;
    /**
     *
     */
    TransportNodeHid.list = hw_transport_node_hid_noevents_1["default"].list;
    /**
     *
     */
    TransportNodeHid.setListenDevicesDebounce = function (delay) {
        listenDevicesDebounce = delay;
    };
    /**
     *
     */
    TransportNodeHid.setListenDevicesPollingSkip = function (conditionToSkip) {
        listenDevicesPollingSkip = conditionToSkip;
    };
    /**
     *
     */
    TransportNodeHid.setListenDevicesDebug = function () {
        console.warn("setListenDevicesDebug is deprecated. Use @ledgerhq/logs instead. No logs will get emitted there anymore.");
    };
    /**
     */
    TransportNodeHid.listen = function (observer) {
        var unsubscribed = false;
        Promise.resolve((0, hw_transport_node_hid_noevents_1.getDevices)()).then(function (devices) {
            var e_1, _a;
            try {
                // this needs to run asynchronously so the subscription is defined during this phase
                for (var devices_2 = __values(devices), devices_2_1 = devices_2.next(); !devices_2_1.done; devices_2_1 = devices_2.next()) {
                    var device = devices_2_1.value;
                    if (!unsubscribed) {
                        var descriptor = device.path;
                        var deviceModel = (0, devices_1.identifyUSBProductId)(device.productId);
                        observer.next({
                            type: "add",
                            descriptor: descriptor,
                            device: device,
                            deviceModel: deviceModel
                        });
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (devices_2_1 && !devices_2_1.done && (_a = devices_2["return"])) _a.call(devices_2);
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
        var _a = (0, listenDevices_1["default"])(listenDevicesDebounce, listenDevicesPollingSkip), events = _a.events, stop = _a.stop;
        var onAdd = function (device) {
            if (unsubscribed || !device)
                return;
            var deviceModel = (0, devices_1.identifyUSBProductId)(device.productId);
            observer.next({
                type: "add",
                descriptor: device.path,
                deviceModel: deviceModel,
                device: device
            });
        };
        var onRemove = function (device) {
            if (unsubscribed || !device)
                return;
            var deviceModel = (0, devices_1.identifyUSBProductId)(device.productId);
            observer.next({
                type: "remove",
                descriptor: device.path,
                deviceModel: deviceModel,
                device: device
            });
        };
        events.on("add", onAdd);
        events.on("remove", onRemove);
        function unsubscribe() {
            unsubscribed = true;
            events.removeListener("add", onAdd);
            events.removeListener("remove", onRemove);
            stop();
        }
        return {
            unsubscribe: unsubscribe
        };
    };
    return TransportNodeHid;
}(hw_transport_node_hid_noevents_1["default"]));
exports["default"] = TransportNodeHid;
